driver/src/driver/stablex_driver.rs

```diff
@@ -271,4 +310,49 @@ mod tests {
                 .was_called_with((batch - 1, Any, Any, Any))
         ));
     }
+
+    #[test]
+    fn test_does_not_submit_solution_for_which_validation_failed() {
+        let contract = StableXContractMock::default();
+        let mut pf = PriceFindingMock::default();
+
+        let orders = vec![create_order_for_test(), create_order_for_test()];
+        let state = create_account_state_with_balance_for(&orders);
+
+        let batch = U256::from(42);
+        contract
+            .get_current_auction_index
+            .given(())
+            .will_return(Ok(batch));
+
+        contract
+            .get_auction_data
+            .given(batch - 1)
+            .will_return(Ok((state.clone(), orders.clone())));
+
+        contract
+            .get_solution_objective_value
+            .given((batch - 1, Val(orders.clone()), Any))
+            .will_return(Err(DriverError::new(
+                "get_solution_objective_value failed",
+                ErrorKind::Unknown,
+            )));
+
+        let solution = Solution {
+            prices: vec![1, 2],
+            executed_sell_amounts: vec![0, 2],
+            executed_buy_amounts: vec![0, 2],
+        };
+        pf.find_prices
+            .given((orders, state))
+            .will_return(Ok(solution));
+
+        let mut driver = StableXDriver::new(&contract, &mut pf);
+        assert!(driver.run().is_err());
+        assert!(!mock_it::verify(
+            contract
+                .submit_solution
+                .was_called_with((batch - 1, Any, Any, Any))
+        ));
```

Comment:
I was thinking there should be some other assertions here... We assert there was an error, and that submit solution was called, but we don't check that it `does_not_submit_solution_for_which_validation_failed`. Am I missing something here?

Reply:
You are missing the ! which negates the statement and this making sure submit_solution was not called.